### ISSUE TITLE
[BUG] azure-storage rename takes over 2 minutes to fail (because the target file already exists) in concurrent context

### ISSUE THREAD

**Issue by findinpath** [Azure/azure-sdk-for-java#39774]
**Describe the bug**
tldr; This issue is likely happening on Azure Storage server side. Naively i'd say that there a backoff mechanism kicking-in on which creates the feeling of `rename` operation simply hanging for several minutes.

Try to rename multiple temporary files on the same time to the same target file `00000000000000000001.json`.

```
2024-04-17T15:17:18.766-0600 BEGIN WRITE 00000000000000000001.json.3d3b414b-ef91-483c-a25f-e970c894badb.tmp
2024-04-17T15:17:18.775-0600 BEGIN WRITE 00000000000000000001.json.b0879e34-1919-4f9d-aeb1-1ec6ecd9dded.tmp
2024-04-17T15:17:18.775-0600 BEGIN WRITE 00000000000000000001.json.4a55e73a-8bf9-43ab-8d48-980b5142f825.tmp
2024-04-17T15:17:19.225-0600 BEFORE RENAME 00000000000000000001.json.b0879e34-1919-4f9d-aeb1-1ec6ecd9dded.tmp
2024-04-17T15:17:19.225-0600 BEFORE RENAME 00000000000000000001.json.4a55e73a-8bf9-43ab-8d48-980b5142f825.tmp
2024-04-17T15:17:19.236-0600 BEFORE RENAME 00000000000000000001.json.3d3b414b-ef91-483c-a25f-e970c894badb.tmp
2024-04-17T15:17:19.681-0600 AFTER RENAME 00000000000000000001.json.b0879e34-1919-4f9d-aeb1-1ec6ecd9dded.tmp
2024-04-17T15:17:19.681-0600 END WRITE 00000000000000000001.json.b0879e34-1919-4f9d-aeb1-1ec6ecd9dded.tmp
2024-04-17T15:17:19.847-0600 AFTER RENAME 00000000000000000001.json.4a55e73a-8bf9-43ab-8d48-980b5142f825.tmp
2024-04-17T15:17:19.847-0600 BEFORE DELETE 00000000000000000001.json.4a55e73a-8bf9-43ab-8d48-980b5142f825.tmp
2024-04-17T15:17:20.277-0600 AFTER DELETE 00000000000000000001.json.4a55e73a-8bf9-43ab-8d48-980b5142f825.tmp
2024-04-17T15:17:20.278-0600 END WRITE 00000000000000000001.json.4a55e73a-8bf9-43ab-8d48-980b5142f825.tmp

2024-04-17T15:19:19.706-0600 AFTER RENAME 00000000000000000001.json.3d3b414b-ef91-483c-a25f-e970c894badb.tmp
2024-04-17T15:19:19.706-0600 BEFORE DELETE 00000000000000000001.json.3d3b414b-ef91-483c-a25f-e970c894badb.tmp
2024-04-17T15:19:20.081-0600 AFTER DELETE 00000000000000000001.json.3d3b414b-ef91-483c-a25f-e970c894badb.tmp
2024-04-17T15:19:20.081-0600 END WRITE 00000000000000000001.json.3d3b414b-ef91-483c-a25f-e970c894badb.tmp
```

As can be seen from the log snippet above `00000000000000000001.json.3d3b414b-ef91-483c-a25f-e970c894badb.tmp` takes **sometimes** over 2 minutes to complete (although unsuccessfuly  because another process managed to perform the rename operation already).

***Exception or Stack Trace***
N/A

**To Reproduce**

Create 3 or more temporary files
Attempt to rename all the temporary files at the same time to the same target file in Azure storage.

***Code Snippet***
This is the original snippet from `trinodb/trino` project:

https://github.com/trinodb/trino/blob/ead5e9c6cb6e6c9e45767330f99cd5ad5babac4a/plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/writer/AzureTransactionLogSynchronizer.java#L43-L75

https://github.com/trinodb/trino/blob/0028b3a4b73a12942c61a69c0c8d33ac2a9be69a/lib/trino-filesystem-azure/src/main/java/io/trino/filesystem/azure/AzureFileSystem.java#L210-L232

The issue I have been working on: https://github.com/trinodb/trino/pull/21560

**Expected behavior**

Renaming should fail fast when it can't be done.

**Screenshots**

**Setup (please complete the following information):**
 - OS: [e.g. iOS] Mac OS Sonoma 14.4.1
 - IDE: [e.g. IntelliJ] IntelliJ
 - Library/Libraries: [e.g. com.azure:azure-core:1.16.0 (groupId:artifactId:version)] azure-core:1.47.0
 - Java version: [e.g. 8] Java 21
 - App Server/Environment: [e.g. Tomcat, WildFly, Azure Function, Apache Spark, Databricks, IDE plugin or anything special]
 - Frameworks: [e.g. Spring Boot, Micronaut, Quarkus, etc] 

If you suspect a dependency version mismatch (e.g. you see `NoClassDefFoundError`, `NoSuchMethodError` or similar), please check out [Troubleshoot dependency version conflict article](https://aka.ms/azsdk/java/dependency/troubleshoot) first. If it doesn't provide solution for the problem, please provide:
- verbose dependency tree (`mvn dependency:tree -Dverbose`)
- exception message, full stack trace, and any available logs

**Additional context**
N/A

**Information Checklist**
Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [x] Bug Description Added
- [x] Repro Steps Added
- [ ] Setup information Added

**Comment by alzimmermsft** [comment]
Thank you for asking this question @findinpath.

I would agree with the "gut check" that this is the retry mechanism kicking in. Do you have any log data or metrics on the response status codes, or exceptions, being thrown when this happens to help troubleshoot this further?

@ibrahimrabab, @ibrandes, FYI

**Comment by findinpath** [comment]
> Do you have any log data or metrics on the response status codes, or exceptions, being thrown when this happens to help troubleshoot this further?

I can make available a self-contained java snipped showcasing the limitation in the next few days if this is necessary.
